### PR TITLE
Enable server API gzip encoding when KONTENA_SERVER_GZIP=true

### PR DESCRIPTION
**Pull request by SpComb**
Use `Rack::Deflater` on the server for JSON API response compression.

Tested together with  #3222 for CLI `Accept-Encoding: gzip` support. Might need some `pre`-release integration testing re `kontena/haproxy`?

### CODE REVIEW

**Comment by kke** [comment]
This has some impact on master CPU usage

**Comment by SpComb** [comment]
> This has some impact on master CPU usage

Yeah, I have not tried to measure/benchmark this in any way.

**Comment by SpComb** [comment]
The error is from the [CLI `LogHelper#buffered_log_json`](https://github.com/kontena/kontena/blob/06a7a06e65a5ef0dada14faee2ae20388b87c1fe/cli/lib/kontena/cli/helpers/log_helper.rb#L78) `rescue` block:

```
DEBUG Excon opts: {:omit_default_port=>true, :connect_timeout=>10, :read_timeout=>30, :write_timeout=>10, :ssl_verify_peer=>true, :middlewares=>[Excon::Middleware::ResponseParser, Excon::Middleware::Expects, Excon::Middleware::Idempotent, Excon::Middleware::Instrumentor, Excon::Middleware::Mock, Excon::Middleware::Decompress], :instrumentor=>Kontena::DebugInstrumentor, :ssl_ca_file=>nil, :ssl_verify_peer_host=>nil}
DEBUG [Request]: GET http://localhost:9292/v1/services/development/null/redis/container_logs?limit=100&follow=1 | Headers: {Accept: application/json, Accept-Encoding: gzip, Authorization: Bearer} 
DEBUG parse log JSON chunk: "\x1F\x8B\b\x00\x84(xZ\x00\x03"
ERROR 743: unexpected token at '' (JSON::ParserError)
/home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/json-2.0.2/lib/json/common.rb:156:in `parse'
/home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/json-2.0.2/lib/json/common.rb:156:in `parse'
/home/kontena/kontena/kontena/cli/lib/kontena/cli/helpers/log_helper.rb:81:in `buffered_log_json'
/home/kontena/kontena/kontena/cli/lib/kontena/cli/helpers/log_helper.rb:46:in `block in stream_logs'
/home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/excon-0.60.0/lib/excon/response.rb:119:in `parse'
/home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/excon-0.60.0/lib/excon/middlewares/response_parser.rb:7:in `response_call'
/home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/excon-0.60.0/lib/excon/connection.rb:414:in `response'
/home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/excon-0.60.0/lib/excon/connection.rb:263:in `request'
/home/kontena/kontena/kontena/cli/lib/kontena/client.rb:329:in `request'
/home/kontena/kontena/kontena/cli/lib/kontena/client.rb:255:in `get_stream'
/home/kontena/kontena/kontena/cli/lib/kontena/cli/helpers/log_helper.rb:58:in `stream_logs'
/home/kontena/kontena/kontena/cli/lib/kontena/cli/helpers/log_helper.rb:21:in `show_logs'
/home/kontena/kontena/kontena/cli/lib/kontena/cli/services/logs_command.rb:20:in `execute'
/home/kontena/kontena/kontena/cli/lib/kontena/command.rb:215:in `run'
/home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/clamp-1.1.2/lib/clamp/subcommand/execution.rb:11:in `execute'
/home/kontena/kontena/kontena/cli/lib/kontena/command.rb:215:in `run'
/home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/clamp-1.1.2/lib/clamp/subcommand/execution.rb:11:in `execute'
/home/kontena/kontena/kontena/cli/lib/kontena/command.rb:215:in `run'
/home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/clamp-1.1.2/lib/clamp/command.rb:132:in `run'
bin/kontena:20:in `<top (required)>'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/cli/exec.rb:75:in `load'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/cli/exec.rb:75:in `kernel_load'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/cli/exec.rb:28:in `run'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/cli.rb:424:in `exec'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/cli.rb:27:in `dispatch'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/cli.rb:18:in `start'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/exe/bundle:30:in `block in <top (required)>'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/friendly_errors.rb:122:in `with_friendly_errors'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/exe/bundle:22:in `<top (required)>'
/home/kontena/.local/bin/bundle:23:in `load'
/home/kontena/.local/bin/bundle:23:in `<main>'
DEBUG parse log JSON chunk: "t\x8FMk\xC30\f\x86\xFF\x8A\xF1q4\xAD\x9D\xB4\xCD\a\xF4\xBA[/c\xA7]\x8Abi%,\x91\x83\xE3\x14J\xE9\x7F\x9F\x92\r\x9Am\xEC\xA8W\xCF\xA3\xD7\xBE\xE9\x06u\xA5w\x90\xA7\xE8\\\x86\xC5\xCE\xD4`\x8C)\xB3\xDA\xD0;\xE9\x95f\xE8H\x88@\xD8\f\x89\x95\xC0y\x8E\xD00\x85\xD3B\x05\xFC\xA9\xE6\xB9\x90\x03\x85K\xE3\xC4\xBE}\xB5 ]\xA8\xF5}G\x1C7<\xB6\xEDf>\xFA\xABC\xDFE\x8C\xE0>\xFE\xD5\x1E\xC2<\t\x7F\x0E\x13\xF8\x17\x7F\x90\xCBP\x04\xF68?\xEB{\xEB|\xA0\xC4\xD8i\xE3\x02A$<A\x94<5\xB6HL*\xABWSV\xE9\xB6\xB2\xC5z\x9Fm\xDF\xE4n\xBC\xF6\x939D\xF4\xE3\xD4\x83\x10Af[\x1D\x95\xB1\xEA\x99j\xB50\xACzR/#s\xC3g\xD5I\xF5A>\xC8\b\xADgZ\xA9\xDE\x87x\xD8gy\xB9\xD6\xF7O\x00\x00\x00\xFF\xFF"
ERROR incompatible character encodings: UTF-8 and ASCII-8BIT (Encoding::CompatibilityError)
/home/kontena/kontena/kontena/cli/lib/kontena/cli/helpers/log_helper.rb:78:in `buffered_log_json'
/home/kontena/kontena/kontena/cli/lib/kontena/cli/helpers/log_helper.rb:46:in `block in stream_logs'
/home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/excon-0.60.0/lib/excon/response.rb:119:in `parse'
/home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/excon-0.60.0/lib/excon/middlewares/response_parser.rb:7:in `response_call'
/home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/excon-0.60.0/lib/excon/connection.rb:414:in `response'
/home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/excon-0.60.0/lib/excon/connection.rb:263:in `request'
/home/kontena/kontena/kontena/cli/lib/kontena/client.rb:329:in `request'
/home/kontena/kontena/kontena/cli/lib/kontena/client.rb:255:in `get_stream'
/home/kontena/kontena/kontena/cli/lib/kontena/cli/helpers/log_helper.rb:58:in `stream_logs'
/home/kontena/kontena/kontena/cli/lib/kontena/cli/helpers/log_helper.rb:21:in `show_logs'
/home/kontena/kontena/kontena/cli/lib/kontena/cli/services/logs_command.rb:20:in `execute'
/home/kontena/kontena/kontena/cli/lib/kontena/command.rb:215:in `run'
/home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/clamp-1.1.2/lib/clamp/subcommand/execution.rb:11:in `execute'
/home/kontena/kontena/kontena/cli/lib/kontena/command.rb:215:in `run'
/home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/clamp-1.1.2/lib/clamp/subcommand/execution.rb:11:in `execute'
/home/kontena/kontena/kontena/cli/lib/kontena/command.rb:215:in `run'
/home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/clamp-1.1.2/lib/clamp/command.rb:132:in `run'
bin/kontena:20:in `<top (required)>'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/cli/exec.rb:75:in `load'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/cli/exec.rb:75:in `kernel_load'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/cli/exec.rb:28:in `run'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/cli.rb:424:in `exec'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/cli.rb:27:in `dispatch'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/cli.rb:18:in `start'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/exe/bundle:30:in `block in <top (required)>'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/friendly_errors.rb:122:in `with_friendly_errors'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/exe/bundle:22:in `<top (required)>'
/home/kontena/.local/bin/bundle:23:in `load'
/home/kontena/.local/bin/bundle:23:in `<main>'
ERROR incompatible character encodings: UTF-8 and ASCII-8BIT (Encoding::CompatibilityError)
/home/kontena/kontena/kontena/cli/lib/kontena/cli/helpers/log_helper.rb:87:in `rescue in buffered_log_json'
/home/kontena/kontena/kontena/cli/lib/kontena/cli/helpers/log_helper.rb:75:in `buffered_log_json'
/home/kontena/kontena/kontena/cli/lib/kontena/cli/helpers/log_helper.rb:46:in `block in stream_logs'
/home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/excon-0.60.0/lib/excon/response.rb:119:in `parse'
/home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/excon-0.60.0/lib/excon/middlewares/response_parser.rb:7:in `response_call'
/home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/excon-0.60.0/lib/excon/connection.rb:414:in `response'
/home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/excon-0.60.0/lib/excon/connection.rb:263:in `request'
/home/kontena/kontena/kontena/cli/lib/kontena/client.rb:329:in `request'
/home/kontena/kontena/kontena/cli/lib/kontena/client.rb:255:in `get_stream'
/home/kontena/kontena/kontena/cli/lib/kontena/cli/helpers/log_helper.rb:58:in `stream_logs'
/home/kontena/kontena/kontena/cli/lib/kontena/cli/helpers/log_helper.rb:21:in `show_logs'
/home/kontena/kontena/kontena/cli/lib/kontena/cli/services/logs_command.rb:20:in `execute'
/home/kontena/kontena/kontena/cli/lib/kontena/command.rb:215:in `run'
/home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/clamp-1.1.2/lib/clamp/subcommand/execution.rb:11:in `execute'
/home/kontena/kontena/kontena/cli/lib/kontena/command.rb:215:in `run'
/home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/clamp-1.1.2/lib/clamp/subcommand/execution.rb:11:in `execute'
/home/kontena/kontena/kontena/cli/lib/kontena/command.rb:215:in `run'
/home/kontena/kontena/kontena/cli/vendor/bundle/ruby/2.3.0/gems/clamp-1.1.2/lib/clamp/command.rb:132:in `run'
bin/kontena:20:in `<top (required)>'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/cli/exec.rb:75:in `load'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/cli/exec.rb:75:in `kernel_load'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/cli/exec.rb:28:in `run'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/cli.rb:424:in `exec'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/cli.rb:27:in `dispatch'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/cli.rb:18:in `start'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/exe/bundle:30:in `block in <top (required)>'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/lib/bundler/friendly_errors.rb:122:in `with_friendly_errors'
/home/kontena/.gem/ruby/2.3.0/gems/bundler-1.16.0/exe/bundle:22:in `<top (required)>'
/home/kontena/.local/bin/bundle:23:in `load'
/home/kontena/.local/bin/bundle:23:in `<main>'
DEBUG [Error]: incompatible character encodings: UTF-8 and ASCII-8BIT (Encoding::CompatibilityError)
incompatible character encodings: UTF-8 and ASCII-8BIT (Encoding::CompatibilityError)
DEBUG Execution took 0.289 seconds
DEBUG SystemExit status 1
```

**Comment by SpComb** [comment]
Streaming commands fixed by not having the CLI send the `Accept-Encoding: gzip` request header for streaming requests: #3242

**Comment by SpComb** [comment]
e2e specs passing together with #3222, i.e. using gzip for all API responses.

Dunno if we have e2e specs for any of the streaming requests like `kontena service logs -f`, though?

File: server/server.rb [hunk 1]
```diff
@@ -15,6 +15,7 @@
 class Server < Roda
   VERSION = File.read('./VERSION').strip
 
+  use Rack::Deflater, include: %w(application/json application/vnd.api+json application/yaml)
```
Comment (kke) [inline server/server.rb]: Of these the master will probably just return `application/json`.

(Maybe the streaming endpoints could set content type to `application/json+streaming` and `Rack::Deflater` would then skip zipping those?)

**Comment by kke** [comment]
This could be env-enablable?

```
use Rack::Deflater, include: %w(application/json) if ENV['KONTENA_ENABLE_GZIP`].to_s == 'true'
```

**Comment by SpComb** [comment]
> This could be env-enablable?

Yes... feature-flagged behind `KONTENA_SERVER_GZIP=true`

**Comment by jakolehm** [comment]
I would not enable this by default in 1.5. So option to enable this via env variable gets my vote.

File: server/server.rb [hunk 2]
```diff
@@ -15,7 +15,7 @@
 class Server < Roda
   VERSION = File.read('./VERSION').strip
 
-  use Rack::Deflater, include: %w(application/json application/vnd.api+json application/yaml)
+  use Rack::Deflater, include: %w(application/json application/vnd.api+json application/yaml) if ENV['KONTENA_SERVER_GZIP'].to_s == 'true'
```
Comment (kke) [inline server/server.rb]: `%w(application/json)` should be enough for master.

**Comment by kke** [comment]
Maybe this could be enabled in the e2e compose file?

**Pull request merged**